cmd/container-disk-v1alpha/main.go

```diff
@@ -0,0 +1,190 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"kubevirt.io/kubevirt/pkg/log"
+)
+
+const (
+	EnvCopyPath            = "COPY_PATH"
+	EnvImagePath           = "IMAGE_PATH"
+	DiskSourceFallbackPath = "/disk"
+	QEMUIMGPath            = "/usr/bin/qemu-img"
+	ReadinessProbeFile     = "/tmp/healthy"
+)
+
+type DiskInfo struct {
+	Format      string `json:"format"`
+	BackingFile string `json:"backing-filename"`
+}
+
+func main() {
+	var err error
+
+	logger := log.DefaultLogger()
+
+	copyPath := os.Getenv(EnvCopyPath)
+	if copyPath == "" {
+		logger.Errorf("Environment Variable %s is not defined.", EnvCopyPath)
+		os.Exit(1)
+	}
+
+	imagePath := os.Getenv(EnvImagePath)
+	imagePath, err = GetImage(imagePath)
+	if err != nil {
+		logger.Reason(err).Errorf("Image lookup failed.")
+		os.Exit(1)
+	}
+
+	imageType, err := GetImageType(imagePath)
+	if err != nil {
+		logger.Reason(err).Error("Could not determine image type.")
+		os.Exit(1)
+	}
+
+	err = CopyImage(imagePath, copyPath, imageType)
+	if err != nil {
+		logger.Reason(err).Errorf("Failed to copy image to the destination.")
+		os.Exit(1)
+	}
+
+	f, err := os.Create(ReadinessProbeFile)
+	if err != nil {
+		logger.Reason(err).Errorf("Failed to mark myself as ready.")
+		os.Exit(1)
+	}
+	f.Close()
+
+	for {
+		time.Sleep(1 * time.Second)
+	}
+}
+
+func GetImage(imagePath string) (string, error) {
+	if imagePath != "" {
+		if _, err := os.Stat(imagePath); err != nil {
+			if os.IsNotExist(err) {
+				return "", fmt.Errorf("No image on path %s", imagePath)
+			} else {
+				return "", fmt.Errorf("Failed to check if an image exists at %s", imagePath)
+			}
+		}
+	} else {
+		files, err := ioutil.ReadDir(DiskSourceFallbackPath)
+		if err != nil {
+			return "", fmt.Errorf("No disk in folder %s", DiskSourceFallbackPath)
+		}
+		if len(files) > 1 {
+			return "", fmt.Errorf("More than one file found in folder %s, only one disk is allowed", DiskSourceFallbackPath)
+		}
+		imagePath = filepath.Join(DiskSourceFallbackPath, files[0].Name())
+	}
+	return imagePath, nil
+}
+
+func GetImageType(imagePath string) (string, error) {
+	var imageExtension = ""
+
+	if strings.HasSuffix(imagePath, ".qcow2") {
```

Comment:
I know you were following the bash script's logic when you created this. The bash script's logic is wrong though. https://github.com/kubevirt/kubevirt/pull/1664#pullrequestreview-194141126

The only reason the bash script appears to work is because of an issue with using a subshell that prevents us from exiting early if the wrong file extension is detected.

we shouldn't be trusting the file extension. Instead we need to be inspecting the image with qemu-img and ensuring that the disk is raw or qcow2. Otherwise, our alpine container disk will fail I believe.

Reply:
:+1: Changed. Still the check is done by the "user"-provided container and this helps to do malicious things if we copy in that container from qcow2 to raw, but we need another check in either virt-handler or virt-launcher too, since only these two components are controlled by us.